### DISCUSSION TITLE
Improve cache-auto-purging logic

### DISCUSSION THREAD

**Discussion by aidenfoxx** [directus/directus#8048] [Feature Requests]
The Directus admin is not cached when using `CACHE_AUTO_PURGE` because navigating anywhere updates the users `last_page` value. This isn't a bug per-se, but it does undermine the value of enabling `CACHE_AUTO_PURGE`.

I'm happy to fix the issue, but first I'd like to understand how would be best to fix it. I have some ideas:

- Skip clearing cache when updating `last_page`
- Only purge items cache when clearing cache from `ItemsService`. Only clear schema cache when clearing cache from `CollectionsService`

The latter feels like the better solution overall.

**Comment by rijkvanzanten** [comment]
> The Directus admin is not cached when using CACHE_AUTO_PURGE because navigating anywhere updates the users last_page value. This isn't a bug per-se, but it does undermine the value of enabling CACHE_AUTO_PURGE.

Good point. This is a very difficult situation to solve for though 🤔 The user could be relationally nested under any other data point, so we can't just ignore user updates in the cache purging. 

The same goes for any other data that was altered. Ideally, we have some sort of way to know where the data point that was updated is cached, and only purge those particular cache records. I'd love to hear any thoughts you might have on how to achieve that

For now, i'll convert this to a feature request, as it's technically working as expected/intended

**Reply by aidenfoxx** [reply]
Yeah. Not purging cache on `last_page` isn't a great solution since that data then becomes unreliable.

My thought was that we could manage the schema caching easily since it's a known quantity. We know which schema keys can exist, so we could on `ItemsService` update:

- Pull schema keys from cache
- Clear cache
- Push schema keys to cache

Likewise, for `CollectionService` changes we can just remove the known schema keys.

The does create the problem of selectively handling cache, and muddies up the logic.

**Reply by rijkvanzanten** [reply]
The schema cache isn't/shouldn't be affected by the cache auto purge in the first place 🤔

**Reply by aidenfoxx** [reply]
Are you sure? The reason I brought this up is that we were seeing extreme hits to our performance to the extent that it's no faster than having cache disabled inside the admin, so I checked our redis instance and there was no schema.

Isn't the problem that you're using the same config for your `schemaCache` and `cache` instances? I mean, they're separate instances, but connecting to the same place. Doing a `flush` on one will affect the other?

Please excuse my ignorance if I'm missing something. I'm not totally familiar with all the code.

**Reply by aidenfoxx** [reply]
If that is the case, the easy fix is just making `schamaCache` memory only.

**Reply by rijkvanzanten** [reply]
> Isn't the problem that you're using the same config for your schemaCache and cache instances? I mean, they're separate instances, but connecting to the same place. Doing a flush on one will affect the other?

This could be, I'll have to double check that. I think they're using different namespaces by default, so I would assume that keyv would only flush the cache based on that namespace, but not 100% sure

**Reply by aidenfoxx** [reply]
As far as I know, redis doesn't have flushing based on namespace natively (don't quote me on that). But looking at the getConfig() code, it doesn't look like they have different namespaces.

Edit: Keyv does support namespacing though, so I guess it does some magic.

**Reply by rijkvanzanten** [reply]
>  it doesn't look like they have different namespaces.

Ahh that would explain your experiences then. You're right, if you use the same storage for the cache, one will flush the other. Lets try adding a custom namespace to the schema one to see if that fixes it

**Reply by aidenfoxx** [reply]
Yeah. Glad we figured it out! I also checked Keyv and it handles the namespacing internally so it should work correctly with redis.

Do you want to handle the fix or shall I create the PR tomorrow?

**Reply by rijkvanzanten** [reply]
> Do you want to handle the fix or shall I create the PR tomorrow?

I'll open a PR, if this is the fix, it should only take a minute 🙂

**Reply by rijkvanzanten** [reply]
All set, done and merged 👍🏻

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
